zephyr/patches/zephyr/i2c-dw-bus-recovery.patch

```diff
@@ -0,0 +1,766 @@
+ 
+ #define DW_IC_STATUS_ACTIVITY_BIT (0)
+ #define DW_IC_STATUS_TFNT_BIT     (1)
+diff --git a/drivers/i2c/Kconfig.dw b/drivers/i2c/Kconfig.dw
+index d666b2aceb2f..549dd20419b0 100644
+--- a/drivers/i2c/Kconfig.dw
++++ b/drivers/i2c/Kconfig.dw
+@@ -28,3 +29,9 @@ config I2C_DW_RW_TIMEOUT_MS
+ 	int "Set the Read/Write timeout in milliseconds"
+ 	depends on I2C_DW
+ 	default 100
++
++config I2C_DW_EXTENDED_SUPPORT
++	bool "Extended DW features"
++	help
++	  This option enables support for the SCL/SDA timeout registers and some
++	  additional features of the DW I2C controller.
+diff --git a/drivers/i2c/i2c_dw.c b/drivers/i2c/i2c_dw.c
+index 1ee30a647dcf..3136cbcd7362 100644
+--- a/drivers/i2c/i2c_dw.c
++++ b/drivers/i2c/i2c_dw.c
+@@ -6,6 +6,10 @@
+  *
+  * SPDX-License-Identifier: Apache-2.0
+  */
++
++#include <zephyr/arch/cpu.h>
++#include <soc.h>
++
+ #include <stddef.h>
+ #include <zephyr/types.h>
+ #include <stdlib.h>
+@@ -16,6 +21,7 @@
+ #include <zephyr/init.h>
+ #include <zephyr/pm/device.h>
+ #include <zephyr/arch/cpu.h>
++#include <zephyr/irq.h>
+ #include <string.h>
+ 
+ #if defined(CONFIG_PINCTRL)
+@@ -53,6 +59,57 @@ static inline uint32_t get_regs(const struct device *dev)
+ 	return (uint32_t)DEVICE_MMIO_GET(dev);
+ }
+ 
++/*
++ * @param dev: DW I2C device instance
++ * @param wrapper_dev: Callers device instance
++ */
++void i2c_dw_register_recover_bus_cb(const struct device *dev, i2c_api_recover_bus_t recover_bus_cb,
++				    const struct device *wrapper_dev)
++{
++	struct i2c_dw_dev_config *const dw = dev->data;
++
++	dw->recover_bus_cb = recover_bus_cb;
++	dw->recover_bus_dev = (struct device *)wrapper_dev;
++}
++
++/* it might call from i2c_transfer api and have already
++ * lock the bus, no need to lock bus again here
++ */
++static int i2c_recovery_bus(const struct device *dev)
++{
++	int ret = 0;
++	struct i2c_dw_dev_config *const dw = dev->data;
++
++	if (dw->recover_bus_cb) {
++		/* callback customize recovery function */
++		ret = dw->recover_bus_cb(dw->recover_bus_dev);
++		if (ret == 0) {
++			/* recovery success */
++			dw->state = I2C_DW_STATE_READY;
++		}
++	}
++	return ret;
++}
++
++int i2c_dw_recovery_bus(const struct device *dev)
++{
++
++	int ret = 0;
++	struct i2c_dw_dev_config *const dw = dev->data;
++
++	/* lock bus */
++	ret = k_sem_take(&dw->bus_sem, K_FOREVER);
++	if (ret != 0) {
++		return ret;
++	}
++	/* do bus recovery */
++	ret = i2c_recovery_bus(dev);
++	/* unlock bus */
++	k_sem_give(&dw->bus_sem);
++
++	return ret;
++}
++
+ #ifdef CONFIG_I2C_DW_LPSS_DMA
+ void i2c_dw_enable_idma(const struct device *dev, bool enable)
+ {
+@@ -346,7 +403,11 @@ static inline void i2c_dw_transfer_complete(const struct device *dev)
+ 
+ 	write_intr_mask(DW_DISABLE_ALL_I2C_INT, reg_base);
+ 	value = read_clr_intr(reg_base);
+-
++#ifdef CONFIG_CPU_CORTEX_M
++	const struct i2c_dw_rom_config *const rom = dev->config;
++	/* clear pending interrupt */
++	NVIC_ClearPendingIRQ(rom->irqnumber);
```

Comment:
Boo - this is a bug in the driver. There should be no references to CMSIS here. It should just be native Zephyr IRQ handling.